src/kubernetes.cc

```diff
@@ -943,21 +943,26 @@ json::value KubernetesReader::GetOwner(
     LOG(DEBUG) << "KindPath returned {" << path_component.first << ", "
                << path_component.second << "}";
 #endif
-    owner = std::move(
-        QueryMaster(path_component.first + "/namespaces/" + ns + "/" +
-                    path_component.second + "/" + name));
-    // Sanity check: because we are looking up by name, the object we get
-    // back might have a different uid.
-    const json::Object* owner_obj = owner->As<json::Object>();
-    const json::Object* metadata = owner_obj->Get<json::Object>("metadata");
-    const std::string owner_uid = metadata->Get<json::String>("uid");
-    if (owner_uid != uid) {
-      LOG(WARNING) << "Owner " << kind << "'" << name << "' (id " << uid
-                   << ") disappeared before we could query it. Found id "
-                   << owner_uid << " instead.";
-      owner.reset();
-      throw QueryException("Owner " + kind + " " + name + " (id " + uid +
-                           ") disappeared");
+    try {
+      owner = std::move(
+          QueryMaster(path_component.first + "/namespaces/" + ns + "/" +
+                      path_component.second + "/" + name));
+      // Sanity check: because we are looking up by name, the object we get
+      // back might have a different uid.
+      const json::Object* owner_obj = owner->As<json::Object>();
+      const json::Object* metadata = owner_obj->Get<json::Object>("metadata");
+      const std::string owner_uid = metadata->Get<json::String>("uid");
+      if (owner_uid != uid) {
+        LOG(WARNING) << "Owner " << kind << "'" << name << "' (id " << uid
+                     << ") disappeared before we could query it. Found id "
+                     << owner_uid << " instead.";
+        owner.reset();
+        throw QueryException("Owner " + kind + " " + name + " (id " + uid +
+                             ") disappeared");
+      }
+    } catch (const QueryException& e) {
+      owners_.erase(encoded_ref);  // Remove the nullptr entry from owners_.
```

Comment:
Instead of erasing the reference on an exception, can we just not add it until we get a success?